File: persona1_java/src/cc4p1/raft/RaftServer.java
package cc4p1.raft;

import cc4p1.common.Frame;
import cc4p1.common.ProtocolMessage;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Acepta conexiones TCP entrantes de:
 *   - otros nodos Raft (RV / AE)          -> peer-to-peer
 *   - el Servidor de Testeo (PROPOSE)     -> cliente que escribe
 *   - el Cliente Vigilante (GET_LOG)      -> cliente que lee
 *
 * Usa un hilo por conexion (a traves de un ExecutorService) para que una
 * conexion lenta no bloquee a las demas, y para poder atender en paralelo
 * a las 3 camaras/servidores de testeo si el proyecto escala a mas nodos.
 */
public class RaftServer implements Runnable {

    private final int port;
    private final RaftNode node;
    private final ExecutorService connectionPool = Executors.newCachedThreadPool();
    private volatile boolean running = true;

    public RaftServer(int port, RaftNode node) {
        this.port = port;
        this.node = node;
    }

    @Override
    public void run() {
        try (ServerSocket serverSocket = new ServerSocket(port)) {
            System.out.println("[RaftServer] Escuchando en el puerto " + port);
            while (running) {
                Socket client = serverSocket.accept();
                connectionPool.submit(() -> handleConnection(client));
            }
        } catch (IOException e) {
            System.err.println("[RaftServer] Error en el socket de escucha: " + e.getMessage());
        }
    }

    public void stop() {
        running = false;
        connectionPool.shutdownNow();
    }

    private void handleConnection(Socket client) {
        try (Socket socket = client) {
            String request = Frame.read(socket.getInputStream());
            String response = dispatch(request);
            Frame.write(socket.getOutputStream(), response);
        } catch (IOException e) {
            // conexion cerrada abruptamente por el otro lado: se ignora
        }
    }

    private String dispatch(String message) {
        String[] f = ProtocolMessage.split(message);
        String type = f[0];

        switch (type) {
            case ProtocolMessage.REQUEST_VOTE: {
                long term = Long.parseLong(f[1]);
                String candidateId = f[2];
                long lastLogIndex = Long.parseLong(f[3]);
                long lastLogTerm = Long.parseLong(f[4]);
                return node.handleRequestVote(term, candidateId, lastLogIndex, lastLogTerm);
            }
            case ProtocolMessage.APPEND_ENTRIES: {
                long term = Long.parseLong(f[1]);
                String leaderId = f[2];
                long prevLogIndex = Long.parseLong(f[3]);
                long prevLogTerm = Long.parseLong(f[4]);
                long leaderCommit = Long.parseLong(f[5]);
                String entries = f.length > 6 ? f[6] : "";
                return node.handleAppendEntries(term, leaderId, prevLogIndex, prevLogTerm, leaderCommit, entries);
            }
            case ProtocolMessage.PROPOSE: {
                int camera = Integer.parseInt(f[1]);
                String clase = f[2];
                String timestamp = f[3];
                String imagePath = f[4];
                return node.handlePropose(camera, clase, timestamp, imagePath);
            }
            case ProtocolMessage.GET_LOG:
                return node.handleGetLog();
            default:
                return "ERROR|tipo_desconocido";
        }
    }
}

File: persona2_python/src/raft/raft_server.py


import socket
import threading

from common import frame, protocol
from raft.raft_node import RaftNode


class RaftServer:
    def __init__(self, port: int, node: RaftNode):
        self.port = port
        self.node = node
        self._server_socket = None
        self._running = True

    def serve_forever(self):
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as server:
            server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            server.bind(("0.0.0.0", self.port))
            server.listen(64)
            self._server_socket = server
            print("[RaftServer] Escuchando en el puerto %d" % self.port)
            while self._running:
                try:
                    conn, _addr = server.accept()
                except OSError:
                    break
                threading.Thread(target=self._handle_connection, args=(conn,), daemon=True).start()

    def stop(self):
        self._running = False
        if self._server_socket:
            try:
                self._server_socket.close()
            except OSError:
                pass

    def _handle_connection(self, conn: socket.socket):
        try:
            with conn:
                request = frame.read_text(conn)
                response = self._dispatch(request)
                frame.write_text(conn, response)
        except (IOError, OSError):
            pass  # conexion cerrada abruptamente por el otro lado: se ignora

    def _dispatch(self, message: str) -> str:
        f = protocol.split(message)
        tipo = f[0]

        if tipo == protocol.REQUEST_VOTE:
            term, candidate_id, last_log_index, last_log_term = int(f[1]), f[2], int(f[3]), int(f[4])
            return self.node.handle_request_vote(term, candidate_id, last_log_index, last_log_term)

        if tipo == protocol.APPEND_ENTRIES:
            term = int(f[1])
            leader_id = f[2]
            prev_log_index = int(f[3])
            prev_log_term = int(f[4])
            leader_commit = int(f[5])
            entries = f[6] if len(f) > 6 else ""
            return self.node.handle_append_entries(term, leader_id, prev_log_index, prev_log_term,
                                                     leader_commit, entries)

        if tipo == protocol.PROPOSE:
            camera = int(f[1])
            clase = f[2]
            timestamp = f[3]
            image_path = f[4]
            return self.node.handle_propose(camera, clase, timestamp, image_path)

        if tipo == protocol.GET_LOG:
            return self.node.handle_get_log()

        return "ERROR|tipo_desconocido"
